### PR TITLE
Block executor configurations, profiling, debugging utility

### PR DESCRIPTION
**Pull request by gelash**
Plumbing and refactoring, some useful for coming changes (e.g. committer backup)

- adds new block executor configurations: 
   - committer back-up behavior: should the workers that commit a transaction perform "backup" of the next txn.  
   - flag for profiling
- adds basic profiling infra for measuring callbacks from the VM to block executor
- propagates more debugging information
- removes X : Executable bounds since that's done differently in the new code cache

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [x] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?
Passes existing tests, new tests for the ViewProfilerState

## Key Areas to Review
Configuration changes are the most sensitive in this PR

### CODE REVIEW

File: aptos-move/aptos-debugger/src/aptos_debugger.rs [hunk 1]
```diff
@@ -430,8 +431,10 @@ fn execute_block_no_limit(
         BlockExecutorConfig {
             local: BlockExecutorLocalConfig {
                 concurrency_level,
-                allow_fallback: true,
+                allow_sequential_block_fallback: true,
                 discard_failed_blocks: false,
+                enable_block_stm_profiling: false,
+                block_stm_committer_backup: BlockSTMCommitterBackup::Default,
```
Comment (gelash) [inline aptos-move/aptos-debugger/src/aptos_debugger.rs:437]: This is e.g. so we can easily disable if needed, but should we also connect it to actual configs (similar to concurrency_level)? @igor-aptos @zekun000

Comment (zekun000) [reply aptos-move/aptos-debugger/src/aptos_debugger.rs:437]: we have execution config here https://github.com/aptos-labs/aptos-core/blob/e292612324aa5a71d398d7c12748a78b6e50dcf1/config/src/config/execution_config.rs#L34

File: aptos-move/block-executor/src/view.rs [hunk 2]
```diff
@@ -1395,37 +1413,70 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> TResourceVi
...
                 if let ReadResult::Metadata(v) = res {
                     v
                 } else {
                     unreachable!("Read result must be Metadata kind")
                 }
-            })
+            });
+
+        if let Some(profiler_state) = &self.maybe_profiler_state {
+            profiler_state.borrow_mut().record_resource_view_stat(
```
Comment (zekun000) [inline aptos-move/block-executor/src/view.rs:1452]: this is pretty ugly lol, we can follow the timer pattern that prometheus has (record on drop)

File: aptos-move/aptos-vm/src/sharded_block_executor/global_executor.rs [hunk 3]
```diff
@@ -62,8 +65,10 @@ impl<S: StateView + Sync + Send + 'static> GlobalExecutor<S> {
             BlockExecutorConfig {
                 local: BlockExecutorLocalConfig {
                     concurrency_level: self.concurrency_level,
-                    allow_fallback: true,
+                    allow_sequential_block_fallback: true,
```
Comment (georgemitenkov) [inline aptos-move/aptos-vm/src/sharded_block_executor/global_executor.rs:68]: Can we have something like fn `aptos_prod_block_executor_config` and have a builder pattern for it defined somewhere so we do not need to set/copy all these and it is also clear what are the prod settings? And for profiling, something along the lines of `fn with_block_stm_profiling(mut self) -> Self`.

**Review by georgemitenkov — COMMENTED** [review]
Quick pass, left a few comments, but in general my main question is if we want the profiler config to be part of the binary. Is it better than a compile-time cfg flag, or similar? Do the newly added checks if profiling is enabled introduce runtime overhead?

File: aptos-move/block-executor/src/profiler.rs [hunk 4]
```diff
@@ -0,0 +1,119 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_logger::info;
+use aptos_mvhashmap::types::{Incarnation, TxnIndex};
+use std::time::{Duration, Instant};
+
+// Total number of callbacks and the time spent, across different (executor) view traits.
+pub(crate) struct ViewProfilerState {
+    pub(crate) resource_view_stats: (Duration, usize),
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/block-executor/src/profiler.rs:10]: oh all this is refactored and beautiful now thanks to @zekun000

File: aptos-move/block-executor/src/view.rs [hunk 5]
```diff
@@ -1719,15 +1842,25 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> TDelayedFie
                     skip,
                 )
             },
+        };
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/block-executor/src/view.rs:1845]: same - refactored this ugly pattern

File: aptos-move/aptos-vm/src/sharded_block_executor/global_executor.rs (hunk 3 above)
Comment (gelash) [reply aptos-move/aptos-vm/src/sharded_block_executor/global_executor.rs:65]: builder pattern sounds good, but I'd hold on until we do some changes that might require further customization, so we can construct the building properly?

File: types/src/block_executor/config.rs [hunk 6]
```diff
@@ -4,16 +4,49 @@
...
 
+/// Configuration for BlockSTM: determines behavior of the workers that rolling
+/// commit transactions and may perform a 'backup' execution of the immediately
+/// following transaction tx in order to make sure the critical path of the block
+/// execution does not contain validation failure and re-execution of tx.
+/// Note: currently, with a VM per worker, it also keeps the VM of the committer
+/// worker warm, and more likely to have a faster back-up execution if the
+/// original execution happens to be slow due to module loading.
+///
+/// The default setting is provided by the implementation whereby certain small
```
[the remark this answers is not in the document]
Comment (gelash) [reply types/src/block_executor/config.rs]: no longer variants, simple bool

File: aptos-move/block-executor/src/profiler.rs [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    pub(crate) fn record_delayed_field_view_stat(&mut self, start: Instant) {
+        let cur = Instant::now();
+        self.delayed_field_view_stats.0 += cur - start;
+        self.delayed_field_view_stats.1 += 1;
+    }
+
+    pub(crate) fn log_info(&self, txn_idx: TxnIndex, incarnation: Incarnation, worker_id: usize) {
+        info!(
+            "TXN = {}, Incarnation = {}, worker_id = {}, callback durations:\
+	     ResourceView total time {:?}, num calls {}; \
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/block-executor/src/profiler.rs]: changed to millis

File: aptos-move/block-executor/src/view.rs (hunk 2 above)
Comment (gelash) [reply aptos-move/block-executor/src/view.rs]: refactored, went from ugly to maybe beautiful :)

File: aptos-move/aptos-debugger/src/aptos_debugger.rs (hunk 1 above)
Comment (gelash) [reply aptos-move/aptos-debugger/src/aptos_debugger.rs]: plumbed the configs as per our discussion (debugger and executor benchmark have options, general config has flag for backup)

File: aptos-move/aptos-vm/src/sharded_block_executor/global_executor.rs (hunk 3 above)
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/sharded_block_executor/global_executor.rs:65]: Sounds good!

File: aptos-move/block-executor/src/view.rs [hunk 8]
```diff
@@ -1515,16 +1542,11 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> TResourceGr
     }
 
     fn is_resource_groups_split_in_change_set_capable(&self) -> bool {
-        match &self.latest_view {
-            ViewState::Sync(_) => true,
-            ViewState::Unsync(_) => true,
-        }
```
Comment (igor-aptos) [inline aptos-move/block-executor/src/view.rs:1521]: doh

File: aptos-move/aptos-vm/src/aptos_vm.rs [hunk 9]
```diff
@@ -2580,9 +2496,12 @@ impl VMExecutor for AptosVM {
...
                 local: BlockExecutorLocalConfig {
-                    concurrency_level: Self::get_concurrency_level(),
-                    allow_fallback: true,
-                    discard_failed_blocks: Self::get_discard_failed_blocks(),
+                    concurrency_level: AptosVMStaticConfig::get_concurrency_level(),
+                    allow_sequential_block_fallback: true,
+                    discard_failed_blocks: AptosVMStaticConfig::get_discard_failed_blocks(),
+                    enable_block_stm_profiling:
+                        AptosVMStaticConfig::get_block_stm_profiling_enabled(),
+                    enable_committer_backup: AptosVMStaticConfig::get_committer_backup_enabled(),
```
Comment (igor-aptos) [inline aptos-move/aptos-vm/src/aptos_vm.rs:2504]: can we just have AptosVMStaticConfig return BlockExecutorLocalConfig itself?

File: aptos-move/aptos-vm/src/aptos_vm.rs [hunk 10]
```diff
@@ -2612,7 +2531,7 @@ impl VMExecutor for AptosVM {
         let ret = sharded_block_executor.execute_block(
             state_view,
             transactions,
-            AptosVM::get_concurrency_level(),
+            AptosVMStaticConfig::get_concurrency_level(),
```
Comment (igor-aptos) [inline aptos-move/aptos-vm/src/aptos_vm.rs:2534]: this should really get BlockExecutorLocalConfig instead.

File: aptos-node/src/utils.rs [hunk 11]
```diff
@@ -55,16 +55,19 @@ pub fn set_aptos_vm_configurations(node_config: &NodeConfig) {
...
-    AptosVM::set_discard_failed_blocks(node_config.execution.discard_failed_blocks);
-    AptosVM::set_num_proof_reading_threads_once(
+    AptosVMStaticConfig::set_concurrency_level_once(effective_concurrency_level as usize);
+    AptosVMStaticConfig::set_committer_backup_enabled_once(
+        node_config.execution.enable_committer_backup,
+    );
+    AptosVMStaticConfig::set_discard_failed_blocks(node_config.execution.discard_failed_blocks);
+    AptosVMStaticConfig::set_num_proof_reading_threads_once(
         node_config.execution.num_proof_reading_threads as usize,
     );
```
Comment (igor-aptos) [inline aptos-node/src/utils.rs:65]: can we just have AptosVMStaticConfig::set_from_execution_configs_once(node_config.execution), which will set all of them?

**Review by igor-aptos — COMMENTED** [review]
I don't understand module executable difference from this PR, but the rest looks good.

if you disable profiling by default, and have identical perf, this sounds good.

**Review by igor-aptos — APPROVED** [review]

**Comment by gelash** [comment]
TODO: add rolling commit timer

**Pull request closed without merging**